### PR TITLE
Use test host in unit tests

### PR DESCRIPTION
**Pull request by ngruson**
Fixes issue #4976

## Changes

@JamesNK gave some recommendations about the usage of `WebApplicationFactory`.
As much as possible, we have implemented that.

Some test cases require a `KestrelServer` implementation for Kestrel metrics to be emitted.
In those cases, a `WebApplication` instance is instantiated to run the unit test against.
Previously, the hardcoded port number 5000 was used to run the host.
The port number is now made variable to isolate the unit test as much as possible.

## Merge requirement checklist

* [X] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (nullable enabled, static analysis, etc.)
* [X] Unit tests added/updated
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [X] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by vishweshbankwar** [comment]
@ngruson - Thank you for taking this. I have not used `MetricCollector` before. But looking at the [link](https://learn.microsoft.com/en-us/aspnet/core/log-mon/metrics/metrics?view=aspnetcore-8.0#test-metrics-in-aspnet-core-apps) shared in #4976, I think it may not be something we can use for testing our scenario. 

`MetricCollector` is testing metrics outside of OTel context by enabling its own `MeterListener`. In our case we want to validate if the user configures OTel SDK and instrumentation as shown below, will they get the desired metrics on `.net8.0` exported?

```csharp
Sdk.CreateMeterProviderBuilder()
   .AddAspNetCoreInstrumentation() 
   .Build();
```
In your test assert on [collector](https://github.com/open-telemetry/opentelemetry-dotnet/blob/56aa1db7d5b95aab8580d291a42eb561e400bb69/test/OpenTelemetry.Instrumentation.AspNetCore.Tests/MetricTests.cs#L150) is independent of assert on [exported Items](https://github.com/open-telemetry/opentelemetry-dotnet/blob/56aa1db7d5b95aab8580d291a42eb561e400bb69/test/OpenTelemetry.Instrumentation.AspNetCore.Tests/MetricTests.cs#L237-L247)

cc: @JamesNK

**Comment by vishweshbankwar** [comment]
regarding : https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/test/OpenTelemetry.Instrumentation.AspNetCore.Tests/InProcServerTests.cs: We should delete this file as the test was added just for e.g. purpose. You could send a separate PR to remove that one.

**Comment by ngruson** [comment]
Other files that could be refactored to use the test host:
  - [`AspNetCoreInstrumentationBenchmarks.cs`](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/test/Benchmarks/Instrumentation/AspNetCoreInstrumentationBenchmarks.cs)
  - [`AspNetCoreInstrumentationNewBenchmarks`](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/test/Benchmarks/Instrumentation/AspNetCoreInstrumentationNewBenchmarks.cs)
  - [`HttpClientInstrumentationBenchmarks`](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/test/Benchmarks/Instrumentation/HttpClientInstrumentationBenchmarks.cs)
  - [`BasicTests.cs`](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs)
  - [`W3CTraceContextTests.cs`](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/test/OpenTelemetry.Instrumentation.W3cTraceContext.Tests/W3CTraceContextTests.cs) (don't know if this can be done in similar fashion. The endpoint is passed to a Python script.)

**Comment by vishweshbankwar** [comment]
> Other files that could be refactored to use the test host:
> 
> * [`AspNetCoreInstrumentationBenchmarks.cs`](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/test/Benchmarks/Instrumentation/AspNetCoreInstrumentationBenchmarks.cs)
> * [`AspNetCoreInstrumentationNewBenchmarks`](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/test/Benchmarks/Instrumentation/AspNetCoreInstrumentationNewBenchmarks.cs)
> * [`HttpClientInstrumentationBenchmarks`](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/test/Benchmarks/Instrumentation/HttpClientInstrumentationBenchmarks.cs)
> * [`BasicTests.cs`](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs)
> * [`W3CTraceContextTests.cs`](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/test/OpenTelemetry.Instrumentation.W3cTraceContext.Tests/W3CTraceContextTests.cs) (don't know if this can be done in similar fashion. The endpoint is passed to a Python script.)

@ngruson Yes, it would be great if you could do that refactor. We don't need to refactor the Benchmarks one for now. BasicTests already uses WebApplicationFactory in most of the tests, we need to refactor the ones not using it like [this](https://github.com/open-telemetry/opentelemetry-dotnet/blob/4ccee25c9f0b35b1900beab98f859c4cd9d713dd/test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs#L1082-L1130) one

**Comment by ngruson** [comment]
Looked a bit more. I don't think we can use `WebApplicationFactory` in the unit tests `ValidateNet8MetricsAsync` and `ValidateNet8RateLimitingMetricsAsync`. 
`WebApplicationFactory` uses `TestServer` under the hood. We need a `KestrelServer` in order to get exported Kestrel metrics.

[KestrelMetrics](https://github.com/dotnet/aspnetcore/blob/350078bbc7b6e6a1fdd67b21fb33092f021771c2/src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelMetrics.cs) is an internal sealed class in aspnetcore so we can easily inject that in a unit test other than using an actual `KestrelServer`. So now I reverted these unit tests mostly back to their original state. A thing I did change was to use a variable port to host the test service instead of hardcoding port 5000. That way it doesn't interfere with other unit tests that are using `WebApplicationFactory`.

Adding a rate limiter to a host created with `WebApplicationFactory` leads to a 404 Not Found response in the unit test. Without the rate limiter, it does return a 200 OK response.
```csharp
var exportedItems = new List<Metric>();

void ConfigureTestServices(IServiceCollection services)
{
    this.meterProvider = Sdk.CreateMeterProviderBuilder()
        .AddAspNetCoreInstrumentation()
        .AddInMemoryExporter(exportedItems)
        .Build();

    services.AddOpenTelemetry()
        .WithMetrics(builder =>
            builder.AddInstrumentation(() => this.meterProvider));

    services.AddRateLimiter(_ => _
        .AddFixedWindowLimiter(policyName: "fixed", options =>
        {
            options.PermitLimit = 4;
            options.Window = TimeSpan.FromSeconds(12);
            options.QueueProcessingOrder = QueueProcessingOrder.OldestFirst;
            options.QueueLimit = 2;
        }));
}

using (var client = this.factory
    .WithWebHostBuilder(builder =>
    {
        builder.ConfigureTestServices(ConfigureTestServices);
        builder.ConfigureLogging(loggingBuilder => loggingBuilder.ClearProviders());
        builder.Configure(app => app.UseRateLimiter());
    })
    .CreateClient())
{
    var response = await client.GetAsync("/api/values").ConfigureAwait(false);
    Assert.True(response.IsSuccessStatusCode); // fails with 404
}
```

How to proceed?

**Comment by vishweshbankwar** [comment]
> Looked a bit more. I don't think we can use `WebApplicationFactory` in the unit tests `ValidateNet8MetricsAsync` and `ValidateNet8RateLimitingMetricsAsync`. `WebApplicationFactory` uses `TestServer` under the hood. We need a `KestrelServer` in order to get exported Kestrel metrics.
> 
> [KestrelMetrics](https://github.com/dotnet/aspnetcore/blob/350078bbc7b6e6a1fdd67b21fb33092f021771c2/src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelMetrics.cs) is an internal sealed class in aspnetcore so we can't easily inject that in a unit test other than using an actual `KestrelServer`. So now I reverted these unit tests mostly back to their original state. A thing I did change was to use a variable port to host the test service instead of hardcoding port 5000. That way it doesn't interfere with other unit tests that are using `WebApplicationFactory`.
> 
> Adding a rate limiter to a host created with `WebApplicationFactory` leads to a 404 Not Found response in the unit test. Without the rate limiter, it does return a 200 OK response.
> 
> ```cs
> var exportedItems = new List<Metric>();
> 
> void ConfigureTestServices(IServiceCollection services)
> {
>     this.meterProvider = Sdk.CreateMeterProviderBuilder()
>         .AddAspNetCoreInstrumentation()
>         .AddInMemoryExporter(exportedItems)
>         .Build();
> 
>     services.AddOpenTelemetry()
>         .WithMetrics(builder =>
>             builder.AddInstrumentation(() => this.meterProvider));
> 
>     services.AddRateLimiter(_ => _
>         .AddFixedWindowLimiter(policyName: "fixed", options =>
>         {
>             options.PermitLimit = 4;
>             options.Window = TimeSpan.FromSeconds(12);
>             options.QueueProcessingOrder = QueueProcessingOrder.OldestFirst;
>             options.QueueLimit = 2;
>         }));
> }
> 
> using (var client = this.factory
>     .WithWebHostBuilder(builder =>
>     {
>         builder.ConfigureTestServices(ConfigureTestServices);
>         builder.ConfigureLogging(loggingBuilder => loggingBuilder.ClearProviders());
>         builder.Configure(app => app.UseRateLimiter());
>     })
>     .CreateClient())
> {
>     var response = await client.GetAsync("/api/values").ConfigureAwait(false);
>     Assert.True(response.IsSuccessStatusCode); // fails with 404
> }
> ```
> 
> How to proceed?

@ngruson Sorry for the delay. I will take a look at this soon. Thanks!

File: test/Directory.Packages.props [hunk 1]
```diff
@@ -2,6 +2,7 @@
   <Import Project="$([MSBuild]::GetPathOfFileAbove(Directory.Packages.props, $(MSBuildThisFileDirectory)..))" />
   <ItemGroup>
     <PackageVersion Update="System.Text.Json" Version="6.0.5" />
+    <PackageVersion Include="Microsoft.Extensions.Telemetry.Testing" Version="8.0.0-rc.2.23510.2" />
```
Comment (vishweshbankwar) [inline test/Directory.Packages.props:5]: This is not needed.

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/OpenTelemetry.Instrumentation.AspNetCore.Tests.csproj [hunk 2]
```diff
@@ -44,4 +45,8 @@
       <CopyToOutputDirectory>Always</CopyToOutputDirectory>
     </EmbeddedResource>
   </ItemGroup>
+
+  <ItemGroup Condition="'$(TargetFramework)' == 'net8.0'">
+    <PackageReference Include="Microsoft.AspNetCore.TestHost" />
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Instrumentation.AspNetCore.Tests/OpenTelemetry.Instrumentation.AspNetCore.Tests.csproj:50]: This is not needed.

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs [hunk 3]
```diff
@@ -1037,71 +1050,92 @@ public async Task DiagnosticSourceExceptionCallBackIsNotReceivedForExceptionsHan
...
         Assert.Equal(0, numberOfExceptionCallbacks);
         Assert.Equal(0, numberOfUnSubscribedEvents);
-        Assert.Equal(2, numberofSubscribedEvents);
-
-        await app.DisposeAsync().ConfigureAwait(false);
+        Assert.Equal(2, numberOfSubscribedEvents);
     }
 
     [Fact]
     public async Task RouteInformationIsNotAddedToRequestsOutsideOfMVC()
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs:1093]: FYI - this test is being removed in https://github.com/open-telemetry/opentelemetry-dotnet/pull/5026

**Comment by vishweshbankwar** [comment]
> > Looked a bit more. I don't think we can use `WebApplicationFactory` in the unit tests `ValidateNet8MetricsAsync` and `ValidateNet8RateLimitingMetricsAsync`. `WebApplicationFactory` uses `TestServer` under the hood. We need a `KestrelServer` in order to get exported Kestrel metrics.
> > [KestrelMetrics](https://github.com/dotnet/aspnetcore/blob/350078bbc7b6e6a1fdd67b21fb33092f021771c2/src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelMetrics.cs) is an internal sealed class in aspnetcore so we can't easily inject that in a unit test other than using an actual `KestrelServer`. So now I reverted these unit tests mostly back to their original state. A thing I did change was to use a variable port to host the test service instead of hardcoding port 5000. That way it doesn't interfere with other unit tests that are using `WebApplicationFactory`.
> > Adding a rate limiter to a host created with `WebApplicationFactory` leads to a 404 Not Found response in the unit test. Without the rate limiter, it does return a 200 OK response.
> > ```cs
> > var exportedItems = new List<Metric>();
> > 
> > void ConfigureTestServices(IServiceCollection services)
> > {
> >     this.meterProvider = Sdk.CreateMeterProviderBuilder()
> >         .AddAspNetCoreInstrumentation()
> >         .AddInMemoryExporter(exportedItems)
> >         .Build();
> > 
> >     services.AddOpenTelemetry()
> >         .WithMetrics(builder =>
> >             builder.AddInstrumentation(() => this.meterProvider));
> > 
> >     services.AddRateLimiter(_ => _
> >         .AddFixedWindowLimiter(policyName: "fixed", options =>
> >         {
> >             options.PermitLimit = 4;
> >             options.Window = TimeSpan.FromSeconds(12);
> >             options.QueueProcessingOrder = QueueProcessingOrder.OldestFirst;
> >             options.QueueLimit = 2;
> >         }));
> > }
> > 
> > using (var client = this.factory
> >     .WithWebHostBuilder(builder =>
> >     {
> >         builder.ConfigureTestServices(ConfigureTestServices);
> >         builder.ConfigureLogging(loggingBuilder => loggingBuilder.ClearProviders());
> >         builder.Configure(app => app.UseRateLimiter());
> >     })
> >     .CreateClient())
> > {
> >     var response = await client.GetAsync("/api/values").ConfigureAwait(false);
> >     Assert.True(response.IsSuccessStatusCode); // fails with 404
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > How to proceed?
> 
> @ngruson Sorry for the delay. I will take a look at this soon. Thanks!

@ngruson Could you please take a look at the errors in build and some comments I left? We can start with your approach.

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs (hunk 3 above)
Comment (ngruson) [reply test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs:1093]: Ok, I didn't change it further now.

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/OpenTelemetry.Instrumentation.AspNetCore.Tests.csproj (hunk 2 above)
Comment (ngruson) [reply test/OpenTelemetry.Instrumentation.AspNetCore.Tests/OpenTelemetry.Instrumentation.AspNetCore.Tests.csproj]: Removed the unneeded package reference.

File: test/Directory.Packages.props (hunk 1 above)
Comment (ngruson) [reply test/Directory.Packages.props]: Removed the unneeded package reference.

**Comment by ngruson** [comment]
I resolved some merge conflicts and now the builds are passing.
Can you see if this is ok?

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs [hunk 4]
```diff
@@ -1046,115 +1032,45 @@ void ConfigureTestServices(IServiceCollection services)
...
 
-            services.AddOpenTelemetry()
-                .WithTracing(builder => builder
-                    .AddInstrumentation(() => this.tracerProvider));
-        }
+        var builder = WebApplication.CreateBuilder();
+        builder.Logging.ClearProviders();
+        var app = builder.Build();
 
-        using (var client = this.factory
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs]: @ngruson Why was this change reverted? This one works fine right?

Comment (ngruson) [reply test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs]: Yes, over the weekend I was struggling to get that one to work properly. 
I just fixed it and now it is working with the use of `WebApplicationFactory'.
Except this particular test is skipped during CI, but maybe that's something for another PR.

Comment (vishweshbankwar) [reply test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs]: you could remove the skip in this PR itself. Also, could you revert the changes in .sln file, unintentional changes i guess.

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs [hunk 5]
```diff
@@ -872,6 +874,10 @@ void ConfigureTestServices(IServiceCollection services)
                         },
                     })
                 .Build();
+
+            services.AddOpenTelemetry()
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs:878]: @ngruson Why is this needed?

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs [hunk 6]
```diff
@@ -942,6 +949,10 @@ void ConfigureTestServices(IServiceCollection services)
...
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs:953]: same as https://github.com/open-telemetry/opentelemetry-dotnet/pull/5040/files#r1402473095

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs [hunk 7]
```diff
@@ -1021,45 +1034,41 @@ public async Task DiagnosticSourceExceptionCallBackIsNotReceivedForExceptionsHan
                 })
                 .Build();
 
-        var builder = WebApplication.CreateBuilder();
-        builder.Logging.ClearProviders();
-        var app = builder.Build();
+            services.AddOpenTelemetry()
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs:1037]: https://github.com/open-telemetry/opentelemetry-dotnet/pull/5040/files#r1402473095

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/MetricTests.cs [hunk 8]
```diff
@@ -117,22 +126,32 @@ public async Task ValidateNet8MetricsAsync()
...
-            .AddInMemoryExporter(metricItems)
-            .Build();
+        void ConfigureTestServices(IServiceCollection services)
+        {
+            this.meterProvider = Sdk.CreateMeterProviderBuilder()
+                .AddAspNetCoreInstrumentation()
+                .AddInMemoryExporter(exportedItems)
+                .Build();
+
+            services.AddOpenTelemetry()
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Instrumentation.AspNetCore.Tests/MetricTests.cs:138]: https://github.com/open-telemetry/opentelemetry-dotnet/pull/5040/files#r1402473095

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/MetricTests.cs [hunk 9]
```diff
@@ -26,55 +26,64 @@
...
-            .Build();
+        var exportedItems = new List<Metric>();
+        void ConfigureTestServices(IServiceCollection services)
+        {
+            this.meterProvider = Sdk.CreateMeterProviderBuilder()
+                .AddAspNetCoreInstrumentation()
+                .AddInMemoryExporter(exportedItems)
+                .Build();
+
+            services.AddOpenTelemetry()
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Instrumentation.AspNetCore.Tests/MetricTests.cs:67]: https://github.com/open-telemetry/opentelemetry-dotnet/pull/5040/files#r1402473095

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/MetricTests.cs [hunk 10]
```diff
@@ -157,27 +179,25 @@ public async Task ValidateNet8RateLimitingMetricsAsync()
...
             .ToArray();
 
-        Assert.Single(activeRequestleasesMetric);
+        Assert.Single(activeRequestLeasesMetric);
         Assert.Single(requestLeaseDurationMetric);
         Assert.Single(limitingRequestsMetric);
 
         // TODO
         // aspnetcore.rate_limiting.request.time_in_queue
         // aspnetcore.rate_limiting.queued_requests
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Instrumentation.AspNetCore.Tests/MetricTests.cs:200]: mising await app.DisposeAsync();

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/MetricTests.cs [hunk 11]
```diff
@@ -157,27 +179,25 @@ public async Task ValidateNet8RateLimitingMetricsAsync()
...
-
-        await app.DisposeAsync();
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Instrumentation.AspNetCore.Tests/MetricTests.cs:180]: need this

**Review by vishweshbankwar — APPROVED** [review]
LGTM: Thanks @ngruson! Please clean up the PR description as well and describe the actual changes.

File: test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs (hunk 5 above)
Comment (ngruson) [reply test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs]: Hi, it doesn't seem to be necessary, I removed it and also looked at your other comments below.
But a seemingly unrelated test is failing now during the code coverage action.

```
Failed OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests.MockCollectorIntegrationTests.TestRecoveryAfterFailedExport [107 ms]
  Error Message:
   System.IO.IOException : Failed to bind to address http://localhost:4317.
```

Comment (vishweshbankwar) [reply test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs]: There is a known issue with `TestRecoveryAfterFailedExport `. Its all clear now.

**Comment by ngruson** [comment]
> LGTM: Thanks @ngruson! Please clean up the PR description as well and describe the actual changes.

Just updated the PR description, is it sufficient? 
You probably don't update the change log for this?

**Comment by vishweshbankwar** [comment]
> > LGTM: Thanks @ngruson! Please clean up the PR description as well and describe the actual changes.
> 
> Just updated the PR description, is it sufficient? You probably don't update the change log for this?

Correct, changelog updates are not done for unit test changes.
